Documentation for CC deployment (#23)
kyguy/strimzi-kafka-operator: documentation/modules/cruise-control/proc-deploying-cruise-control.adoc

// This module is included in the following assemblies:
//
// assembly-cruise-control-concepts.adoc

[id='proc-deploying-cruise-control-{context}']
= Deploying Cruise Control

> Deploy Cruise Control through the `cruiseControl` property in the `Kafka` resource.

Reviewer: ```suggestion
Deploy Cruise Control to your {ProductName} cluster through the `cruiseControl` property in the `Kafka` resource.
```

> Deploy Cruise Control through the `Kafka` resource in the `cruiseControl` properties.

Reviewer 2: This is the wrong way round. We deploy CC throught the `cruiseControl` property of the `Kafka` resource?

Reviewer 2: Also is "property" the correct term? Is it not a configuration?

Author: I think "property" is the correct term, it's not a k8s specific term but one we seem to use in our docs to refer to the fields/sections of a resource, I am happy to use "field" too. It is a configuration but I am worried it might get confused with the `cruiseControl.config` section specifically

Refer to the xref:ref-sample-kafka-resource-config-deployment-configuration-kafka[sample Kafka YAML configuration] for an overview of the `Kafka` resource and its properties.

> The properties relevant to the Cruise Control deployment are shown in this procedure.

Reviewer: ^ I think you can delete line 12.

> You can configure these properties as part of a deployment or redeployment of the Kafka cluster.

Reviewer: ```suggestion
You can configure `cruiseControl` properties as part of a deployment or redeployment of a Kafka cluster.
```

.Prerequisites

* A Kubernetes cluster
* A running Cluster Operator

.Procedure

> . Edit the `cruiseControl` properties for the `Kafka` resource.

Reviewer 2: Aren't we editing the `cruiseControl` "property" singular? Makes more sense to say "edit the configuration options under the `cruiseControl` property of the `kafka` resource. 

Also I am not sure "property" is the correct word (unless this is K8s specific term).

Author: I updated to the singular "property". There are two things I am trying to keep distinct: the Cruise Control Deployment configuration, `Kafka.cruiseControl`, and the Cruise Control application configuration, `Kafka.cruiseControl.config`. I would like to refer to `Kafka.cruiseControl.config` as a property for Cruise Control configuration

+
The properties you can configure are shown in this example configuration:
+
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-cluster
spec:
  # ...
  cruiseControl:
    replicas: 1 <1>
    config: <2>

>       default.goals: >

Reviewer: Just checking that the `>` in line 38 is meant to be there.

Also, should we explain what the _default goals_ are?

Author: The `>` is just for formatting a multiline yaml string, included to help make the string of goals more readable. I don't think we need to explain any Cruise Control specific configuration settings like `default.goals`, I think it's enough to redirect users to the Cruise Control docs which explain what these are

Reviewer: Thanks for the explanation. OK. We can always add more info at a later date.

>          com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,

Reviewer 2: If this is a list then each line should start with a "-" and there should not be "," at the end of each line

Author: Not meant to be a yaml list, just a multiline yaml string, (too make it easier to read). It originally looked like this: 
```
cruiseControl:    
  config:
    default.goals: com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
    cpu.balance.threshold: 1.1
    ...
```

Reviewer 2: Ok but this should probably be a YAML list to make it easier to read and write for a user. Would this require operator side parsing?

Author: It would require some additional parsing logic in the operator, but it can be done.  I can update it in the docs here after I implement the parsing logic

         com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
      cpu.balance.threshold: 1.1
      metadata.max.age.ms: 300000
      send.buffer.bytes: 131072

>       ...

Reviewer: ```suggestion
      # ...
```

    resources: <3>
      requests:
        cpu: 200m
        memory: 64Mi
      limits:
        cpu: 500m
        memory: 128Mi
    logging: debug <4>
    template: <5>
      pod:
        metadata:
          labels:
            label1: value1
        securityContext:
          runAsUser: 1000001
          fsGroup: 0
        terminationGracePeriodSeconds: 120
    readinessProbe: <6>
      initialDelaySeconds: 15
      timeoutSeconds: 5
    livenessProbe: <7>
      initialDelaySeconds: 15
      timeoutSeconds: 5
# ...
----

> <1> Number of application instances, can be set to 1 to run or 0 to pause the application.

Reviewer: ```suggestion
<1> Number of Cruise Control application instances; set to `1` to run or `0` to pause the application.
```

> <2> Config specifies the Cruise Control configuration. xref:ref-cruise-control-configuration-{context}[Standard Cruise Control configuration may be provided, restricted to those properties not managed directly by {ProductName}].

Reviewer: ```suggestion
<2> Specifies the Cruise Control configuration. You can provide any xref:ref-cruise-control-configuration-{context}[standard configuration option] apart from those managed directly by {ProductName}.
```

> <3> xref:assembly-resource-limits-and-requests-deployment-configuration-kafka[CPU and memory resources to reserve].

Reviewer: ```suggestion
<3> CPU and memory resources reserved for Cruise Control. For more information, see xref:assembly-resource-limits-and-requests-deployment-configuration-kafka[].
```

<4> Logging configuration, to log messages with a given severity (debug, info, warn, error, fatal) or above.
<5> xref:assembly-customizing-deployments-str[Customization of deployment templates and pods].
<6> xref:assembly-healthchecks-deployment-configuration-kafka[Healthcheck readiness probes].
<7> xref:assembly-healthchecks-deployment-configuration-kafka[Healthcheck liveness probes].

. Create or update the resource:
+
[source,shell,subs="+quotes"]
----
kubectl apply -f _kafka.yaml_
----

> 

Reviewer: Please add a verification step.

Author: What do you mean by verification step? If the yaml is invalid, k8s will prevent it from being applied and print an error message.

Reviewer: I mean a step to verify that Cruise Control was successfully deployed. Something like:

`kubectl get <type> --selector app=strimzi`

.What to do next

> After configuring and deploying Cruise Control, you can create a xref:proc-cruise-control-create-rebalance-resource[create a KafkaRebalance resource]

Reviewer: ```suggestion
After configuring and deploying Cruise Control, you can create a xref:proc-cruise-control-create-rebalance-resource[create a `KafkaRebalance` resource].
```

Reviewer: Add "TBC" to indicate the module has yet to be created.
